router/diag_cntl.c

```diff
@@ -556,6 +583,114 @@ void diag_cntl_set_buffering_mode(struct peripheral *perif, int mode)
 	}
 }
 
+struct list_head *diag_get_diagid_list_head(void)
+{
+	return &diag_id_list;
+}
+
+int diag_add_diag_id_to_list(uint8_t diag_id, const char *process_name, struct peripheral *perif)
+{
+	struct diag_id_tbl_t *new_diag_id = NULL;
+
+	if (!process_name || diag_id == 0)
+		return -EINVAL;
+
+	new_diag_id = malloc(sizeof(struct diag_id_tbl_t));
+	if (!new_diag_id)
+		return -ENOMEM;
+
+	new_diag_id->diag_id = diag_id;
+	memset(new_diag_id->process_name, '\0', sizeof(new_diag_id->process_name));
+	(void)strncpy(new_diag_id->process_name, process_name, sizeof(new_diag_id->process_name) -1);
+	list_add(&diag_id_list, &new_diag_id->node);
+
+	return 0;
+}
+
+int diag_get_diag_id(const char *name, uint32_t *diag_id)
+{
+	struct diag_id_tbl_t *diag_id_item = NULL;
+
+	if (!name || !diag_id)
+		return -EINVAL;
+
+	list_for_each_entry(diag_id_item, &diag_id_list, node) {
+		if (!strcmp(diag_id_item->process_name, name)) {
+			*diag_id = diag_id_item->diag_id;
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+int diag_check_duplicate_diag_id(uint32_t diag_id)
+{
+	struct diag_id_tbl_t *diag_id_item = NULL;
+
+	list_for_each_entry(diag_id_item, &diag_id_list, node) {
+		if (diag_id == diag_id_item->diag_id)
+			return 1;
+	}
+
+	return 0;
+}
+
+static int diag_cntl_process_diag_id(struct peripheral *peripheral, struct diag_cntl_hdr *hdr, size_t len)
+{
+	struct diag_cntl_cmd_diag_id_v2 *pkt_v2 = to_cmd_diag_id_v2(hdr);
+	struct diag_cntl_cmd_diag_id *pkt = to_cmd_diagid(hdr);
+	struct diag_cntl_cmd_diag_id rsp = {0};
+	uint32_t version = 0, local_diag_id = 0;
+	static uint32_t diag_id = DIAG_ID_APPS;
+	char *process_name = NULL;
+	int ret;
+
+	if (pkt == NULL || len == 0)
+		return -EINVAL;
+
+	version = pkt->version;
+	if (peripheral->support_diagid_v2 && version >= DIAG_ID_VERSION_2) {
+		if (len < (sizeof(struct diag_cntl_cmd_diag_id_v2 ) -
+			(MAX_DIAGID_STR_LEN - MIN_DIAGID_STR_LEN))) {
+			return -EINVAL;
+		}
+		process_name = (char *)&pkt_v2->feature_mask + pkt_v2->feature_len;
+	} else {
+		if (len < (sizeof(struct diag_cntl_cmd_diag_id) -
+			(MAX_DIAGID_STR_LEN - MIN_DIAGID_STR_LEN))) {
+			return -EINVAL;
+		}
+		process_name = (char*)&pkt->process_name;
+	}
+
+	ret = diag_get_diag_id(process_name, &local_diag_id);
+	if (!ret) {
+		if (version == DIAG_ID_VERSION_3)
```

Comment:
this should be version >= DIAG_ID_VERSION_3

Reply:
thanks for you review, I have modified it.